src/Stryker.Core/Stryker.Core/InjectedHelpers/Coverage/CommunicationChannel.cs

```diff
@@ -0,0 +1,163 @@
+        private readonly object _lck = new object();
+
+        private byte[] _buffer;
+        private int _cursor;
+        private bool _processingHeader;
+        private bool _started;
+
+        public event MessageReceived RaiseReceivedMessage;
+
+        public CommunicationChannel(PipeStream stream, string name)
+        {
+            _pipeName = name;
+            _pipeStream = stream;
+        }
+
+        public static CommunicationChannel Client(string pipeName, int timeout = -1)
+        {
+            var pipe = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut,
+                PipeOptions.Asynchronous);
+            try
+            {
+                pipe.Connect(timeout);
+            }
+            catch (TimeoutException)
+            {
+                pipe.Dispose();
+                throw;
+            }
+            return new CommunicationChannel(pipe, $"{pipeName}:C");
+        }
+
+        public void Start()
+        {
+            lock(_lck)
+            {
+                if (_started)
+                {
+                    return;
+                }
+                _started = true;
+            }
+
+            Begin();
+        }
+
+        private void Begin(bool init = true)
+        {
+            if (init)
+            {
+                if (_buffer != null && !_processingHeader)
+                {
+                    var message = Encoding.Unicode.GetString(_buffer);
+                    Log($"Received message: [{message}] ({_buffer.Length} bytes).");
+                    RaiseReceivedMessage?.Invoke(this, message);
+                }
+                _processingHeader = !_processingHeader;
+                var len = _processingHeader ? 4 : BitConverter.ToInt32(_buffer,0);
+                if (len<0)
+                {
+                    Log($"Got invalid length, synchro lost. Aborting!");
+                    _pipeStream.Close();
+                    return;
+                }
+                _buffer = new byte[len];
+                _cursor = 0;
+                if (!_processingHeader && _buffer.Length == 0)
+                {
+                    // we have NO DATA to read, notify of empty message and wait to read again.
+                    Log("Empty message.");
+                    Begin();
+                    return;
+                }
+            }
+
+            try
+            {
+                var bufferLength = _buffer.Length-_cursor;
+                Log($"Begin Read {bufferLength} bytes.");
+                _pipeStream.BeginRead(_buffer, _cursor, bufferLength, WhenReceived, null);
+            }
+            catch (ObjectDisposedException)
+            {
+                Log($"Nothing to read, connection closed.");
+            }
+            catch (IOException e)
+            {
+                Log($"Begin Read {e} exception.");
+            }
+        }
+
+        private void Log(string message)
+        {
+            // TODO: control this with logging options
```

Comment:
overall redesign of the logging infrastructure. It has been improved. I think it is good enough for current usage  (as in: you have debug info when you want them and no noise when you don't). What is missing is finer grained control, but it would require the usage of other environment variables and I am reluctant to add more